invokeai/app/invocations/grounded_sam.py

```diff
@@ -0,0 +1,203 @@
+from pathlib import Path
+from typing import Literal
+
+import numpy as np
+import numpy.typing as npt
+import torch
+from PIL import Image
+from transformers import AutoModelForMaskGeneration, AutoProcessor, pipeline
+from transformers.models.sam import SamModel
+from transformers.models.sam.processing_sam import SamProcessor
+from transformers.pipelines import ZeroShotObjectDetectionPipeline
+
+from invokeai.app.invocations.baseinvocation import BaseInvocation, invocation
+from invokeai.app.invocations.fields import ImageField, InputField
+from invokeai.app.invocations.primitives import ImageOutput
+from invokeai.app.services.shared.invocation_context import InvocationContext
+from invokeai.backend.grounded_sam.detection_result import DetectionResult
+from invokeai.backend.grounded_sam.grounding_dino_pipeline import GroundingDinoPipeline
+from invokeai.backend.grounded_sam.mask_refinement import mask_to_polygon, polygon_to_mask
+from invokeai.backend.grounded_sam.segment_anything_model import SegmentAnythingModel
+
+GROUNDING_DINO_MODEL_ID = "IDEA-Research/grounding-dino-tiny"
+SEGMENT_ANYTHING_MODEL_ID = "facebook/sam-vit-base"
```

Comment:
There are different GroundingDINO and SAM models out there. Do we want to lock in to these? If the model license for the others permit, maybe we should just drop them in as options for the user to pick.

Reply:
@hipsterusername

Reply:
Grounding DINO Options (all Apache 2.0):
- `IDEA-Research/grounding-dino-tiny`, 689MB
- `IDEA-Research/grounding-dino-base`, 933MB

SAM Options (all Apache 2.0):
- `facebook/sam-vit-base`, 375MB
- `facebook/sam-vit-large`, 1.25GB
- `facebook/sam-vit-huge`, 2.56GB

So, licenses are all ok. I'll do a quick test to see if there is a qualitative performance difference. The only downside I can think of to exposing all the options is that we don't offer an easy way for users to clean them up to save disk space (the user would have to find the hidden download cache directory and manually delete the files).

Reply:
Because we are using Invoke's Model Manager API, the models get downloaded to the `root > models > .download_cache` folder right? So if a user needs to manually delete, we can guide them there. But also, like you said this needs to be addressed via the client but that's an issue that needs expanding on the Model Manager end.

As for this PR, I am of the opinion that we need to expose the options and let the user decide. We can restrict to the best in the Linear UI but let the workflow users have their share of flexibility.

Reply:
> Because we are using Invoke's Model Manager API, the models get downloaded to the root > models > .download_cache folder right? 

Correct. Ok, I'll add them all.

Reply:
Done [27ac61a](https://github.com/invoke-ai/InvokeAI/pull/6701/commits/27ac61a4fb8ea7fb5a6f3482473ddf080fbb363f)